README.md
<div align="center">

<img src="icon.png" width="128" alt="NFS Free Edition"/>

# Network File System (NFS) — Free Edition

**Free, open-source aaPanel plugin for complete NFS management**

Share directories, mount remote filesystems, monitor your NFS server<br/>
and diagnose mount failures — all from a graphical panel interface.

<br/>

[![License](https://img.shields.io/github/license/jalexiscv/aaPanel-nfs-free?color=blue&style=flat-square)](LICENSE)
[![Version](https://img.shields.io/badge/version-1.1-brightgreen?style=flat-square)](https://github.com/jalexiscv/aaPanel-nfs-free/releases)
[![Python](https://img.shields.io/badge/python-3.6%2B-3776ab?style=flat-square&logo=python&logoColor=white)](https://www.python.org/)
[![Platform](https://img.shields.io/badge/platform-Linux-FCC624?style=flat-square&logo=linux&logoColor=black)](https://github.com/jalexiscv/aaPanel-nfs-free)
[![aaPanel](https://img.shields.io/badge/aaPanel-plugin-4A90D9?style=flat-square)](https://www.aapanel.com/)
[![NFS](https://img.shields.io/badge/NFS-v3%20%7C%20v4-informational?style=flat-square)](README_EN.md#1-about-the-nfs-protocol)

<br/>

[![Stars](https://img.shields.io/github/stars/jalexiscv/aaPanel-nfs-free?style=social)](https://github.com/jalexiscv/aaPanel-nfs-free/stargazers)
[![Forks](https://img.shields.io/github/forks/jalexiscv/aaPanel-nfs-free?style=social)](https://github.com/jalexiscv/aaPanel-nfs-free/network/members)
[![Issues](https://img.shields.io/github/issues/jalexiscv/aaPanel-nfs-free?style=social)](https://github.com/jalexiscv/aaPanel-nfs-free/issues)

</div>

---

## Features

| | |
|---|---|
| 📂 **NFS Shares** | Create and manage exports with per-client IP/CIDR control, rw/ro modes and squash policies |
| 🔗 **Remote Mounts** | Mount remote NFS paths with full protocol options: NFSv3/v4, rsize/wsize, TCP/UDP, hard/soft |
| 📊 **Server Monitoring** | Live status of `nfs-server` and `rpcbind`, RPC service list, `/proc/net/rpc` stats, `nfsiostat` I/O metrics |
| 🚀 **Auto-mount at Boot** | SysV init service mounts all `auto_mount=1` entries automatically on system startup |
| 🔍 **Intelligent Diagnostics** | 7-step failure analysis — ping, IP detection, port check, showmount, CIDR matching — with actionable suggestions |
| 📋 **Activity Log** | Thread-safe audit log of every operation, filterable by event type, result and IP address |
| 🔄 **Connection Tracking** | Detects mounts and client connections that happen outside the plugin in real time |

---

## Quick Install

```bash
# 1. Copy plugin files to the aaPanel plugin directory
cp -r nfs_free/ /www/server/panel/plugin/nfs_free/

# 2. Run the installer
sudo bash /www/server/panel/plugin/nfs_free/install.sh install
```

The installer handles NFS packages, fixed-port configuration for mountd (20048),
firewall rules, SysV init service and systemd service activation automatically.

---

## Screenshots

### Dashboard

![NFS Free Edition — Dashboard](assets/Snap-2023-08-29T11:20:00-05:00.png)

### Shared List

![NFS Free Edition — Shared List](assets/nfs-shared-list.png)

### Mount List

![NFS Free Edition — Mount List](assets/nfs-mount-list.png)

### Service Status

![NFS Free Edition — Service Status](assets/nfs-service-status.png)

### Activity Log

![NFS Free Edition — Activity Log](assets/nfs-activity-log.png)

---

## Ports Required on the Server

| Port  | Service | Protocol | Role |
|-------|---------|----------|------|
| 111   | rpcbind | TCP/UDP  | Port mapper |
| 2049  | nfs     | TCP      | NFS data transfer |
| 20048 | mountd  | TCP/UDP  | Mount requests (pinned) |
| 32874 | lockd   | TCP/UDP  | File locking |
| 32876 | statd   | TCP/UDP  | State recovery |

---

## Documentation

<div align="center">

| Language | File |
|----------|------|
| 🇪🇸 Español | [README_ES.md](README_ES.md) |
| 🇧🇷 Português | [README_PT.md](README_PT.md) |
| 🇬🇧 English | [README_EN.md](README_EN.md) |
| 🇯🇵 日本語 | [README_JA.md](README_JA.md) |
| 🇩🇪 Deutsch | [README_DE.md](README_DE.md) |
| 🇷🇺 Русский | [README_RU.md](README_RU.md) |

</div>

---

## Contributing

1. Fork the repository
2. Create a feature branch: `git checkout -b feature/my-feature`
3. Commit your changes: `git commit -m 'Add: my feature'`
4. Push and open a Pull Request

See [README_EN.md](README_EN.md#12-contributing) for full contribution guidelines.

---

<div align="center">

Made with care by **[Jose Alexis Correa Valencia](https://github.com/jalexiscv)** · Colombia 🇨🇴

[![GitHub](https://img.shields.io/badge/GitHub-jalexiscv-181717?style=flat-square&logo=github)](https://github.com/jalexiscv)
[![LinkedIn](https://img.shields.io/badge/LinkedIn-jalexiscv-0A66C2?style=flat-square&logo=linkedin)](https://www.linkedin.com/in/jalexiscv/)
[![Email](https://img.shields.io/badge/Email-jalexiscv%40gmail.com-EA4335?style=flat-square&logo=gmail)](mailto:jalexiscv@gmail.com)

*Network File System (NFS) Free Edition — Copyright © 2023 Jose Alexis Correa Valencia — Published July 31, 2023 — MIT License*

</div>


CHANGELOG.md
# Changelog

All notable changes to **NFS Free Edition** will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

---

## [Unreleased]

---

## [1.1] - 2023-10-23

### Added

#### Backend — `nfs_free_main.py`
- `check_update` — queries the GitHub Releases API (`/repos/jalexiscv/aaPanel-nfs-free/releases/latest`) to compare the installed version against the latest published release; result cached for 1 hour in `config/update_cache.json` to respect GitHub's unauthenticated rate limit

#### Frontend — `index.html`
- Update banner in the Overview tab: shown automatically when `check_update` reports a newer version; displays current vs. latest version numbers and a direct link to the GitHub release page

#### Repository infrastructure
- `info.json` — `home` field updated to point to the GitHub Releases page instead of the author profile
- `.gitignore` — added `.claude/` to prevent session data from being committed
- `assets/` — real UI screenshots added for all five plugin tabs (Overview, Shared list, Mount list, Service status, Activity log)
- `README.md` — ASCII UI mockups replaced with actual screenshots; plugin icon updated to a server-rack and network-tree design at 512×512 px; icon display size bumped to 128 px in the banner
- `README_*.md` — icon dimension annotation corrected from 48×48 to 512×512 in all six language files
- `.github/workflows/release.yml` — release step now extracts the matching `## [x.y]` section from `CHANGELOG.md` via `awk` and passes it as `--notes-file` instead of using `--generate-notes`

---

## [1.0] - 2023-07-31

### Added

#### Backend — `nfs_free_main.py`
- `create_share` / `modify_share` / `remove_share` — full CRUD for NFS exports; writes `/etc/exports` and reloads via `exportfs -ra` on every change
- `get_share_list` / `get_share_find` — list and lookup configured shares
- `show_ip_share_list` — query exports available on a remote server via `showmount -e`
- `create_mount` / `modify_mount` / `remove_mount` — full CRUD for remote NFS mounts with all protocol options (NFSv3/v4, rsize/wsize, timeo, retrans, hard/soft, TCP/UDP, noresvport)
- `to_mount` / `to_umount` — execute mount and unmount operations with live status
- `auto_mount` — boot-time service that mounts all entries with `auto_mount=1`
- `get_mount_list` — enriches configured mounts with live kernel state (`/proc/mounts`) and I/O metrics
- `get_server_status` — checks `nfs-server` and `rpcbind` via `systemctl is-active` and returns registered RPC services via `rpcinfo -p`
- `server_admin` — start / stop / restart / reload NFS services
- `get_overview` — aggregates all dashboard data in a single call (server health, share count, mount status, I/O totals, recent log)
- `get_nfsstat` — parses `/proc/net/rpc/nfsd` and `/proc/net/rpc/nfs` into structured NFSv3/v4 operation counters
- `get_nfsiostat` / `_parse_nfsiostat` — parses `nfsiostat` output into per-mount read/write metrics (ops/s, KB/s, retransmissions, avg RTT/exe)
- `get_connections` — returns active outgoing mounts and incoming clients via `showmount -a`
- `_scan_connection_changes` — detects mount/unmount events that occurred outside the plugin and logs them automatically
- `get_disk_mounts` — lists all NFS mounts active in the kernel from `/proc/mounts`
- `_diagnose_mount_failure` — 7-step intelligent diagnostics: reachability (ping), outgoing IP detection (`ip route get`), port checks (111/2049/20048 via `/dev/tcp`), export enumeration (`showmount -e`), path validation, CIDR-aware IP access control matching, and kernel error keyword analysis
- `fix_mountd_port` / `_ensure_mountd_port` — pins mountd to port 20048, statd to 32876 and lockd to 32874 in `/etc/nfs.conf` for predictable firewall rules
- `_log` — thread-safe activity log using `fcntl.LOCK_EX`; stores up to 1000 entries in `config/activity.log.json`
- `get_log` / `clear_log` — retrieve log with filters (event, result, IP) and truncate

#### Service bootstrap — `nfs_free_service`
- Lightweight Python entry point executed at boot by the SysV init service
- Configures Python module paths for the panel environment and calls `auto_mount()`
- Silent fail on exception to avoid blocking system startup

#### Scripts
- `install.sh` — installs NFS system packages (apt / yum / dnf), pins daemon ports, registers `/etc/init.d/nfs_free`, enables systemd services, sets plugin permissions and signals panel reload; supports `install` and `uninstall` subcommands
- `init.sh` — SysV-compatible init script (`/etc/init.d/nfs_free`); supports `start`, `stop`, `restart`, `reload`, `status`; auto-detects aaPanel Python virtualenv
- `repair.sh` — re-applies fixed-port configuration and restarts the init service; use when mountd reverts to a random port after a reboot or package upgrade
- `upgrade.sh` — upgrades NFS system packages and re-applies port configuration

#### Frontend — `index.html`
- **Overview tab** — dashboard with four metric cards (shares, mounts, disk usage, I/O ops), live service status indicators, last 8 activity log entries, and quick-action buttons
- **Shared list tab** — table of NFS exports with inline add / edit / delete; port requirements panel with security warning
- **Mount list tab** — table of remote mounts with live mounted/unmounted status badge, per-mount I/O indicators (ops/s, read KB/s, write KB/s, retransmissions) and detection of untracked mounts
- **Service status tab** — animated service cards for `nfs-server` and `rpcbind` with start / stop / restart / reload controls; filterable RPC service chips; NFSv3/v4 statistics tabs sourced from `/proc/net/rpc`
- **Activity log tab** — full audit log with filters by event type, result and IP, colour-coded event badges and configurable auto-refresh

#### Documentation

init.sh
#!/bin/bash
# =============================================================================
#  Network File System (NFS) — Free Edition
#  SysV Init Script — Auto-mount service
# =============================================================================
#  Description : SysV-compatible init script installed at /etc/init.d/nfs_free.
#                Runs at system boot (runlevels 2–5) to auto-mount all NFS
#                shares that have auto_mount=1 in the plugin configuration.
#                Uses the aaPanel Python environment when available, falling
#                back to the system Python 3 interpreter.
#
#  Usage       : /etc/init.d/nfs_free {start|stop|restart|reload|status}
#
#  Runlevels   : start on 2 3 4 5  (priority 55)
#                stop  on 0 1 6    (priority 25)
#
#  Author      : Jose Alexis Correa Valencia (jalexiscv)
#  Email       : jalexiscv@gmail.com
#  GitHub      : https://github.com/jalexiscv
#  LinkedIn    : https://www.linkedin.com/in/jalexiscv/
#  Version     : 1.0
#  License     : MIT — Free Software
#  Date        : July 14, 2023
#  Copyright   : © 2023 Jose Alexis Correa Valencia
# =============================================================================

# chkconfig: 2345 55 25

### BEGIN INIT INFO
# Provides:          nfs_free
# Required-Start:    $all
# Required-Stop:     $all
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: NFS Free — Auto-mount service
# Description:       Auto-mounts NFS shares configured in nfs_free plugin
### END INIT INFO

PANEL_PATH=/www/server/panel
PLUGIN_PATH=$PANEL_PATH/plugin/nfs_free
SERVICE_SCRIPT=$PLUGIN_PATH/nfs_free_service
cd $PANEL_PATH

# Prefer the panel's virtualenv; fall back to system Python 3
if [ -f $PANEL_PATH/pyenv/bin/activate ]; then
    source $PANEL_PATH/pyenv/bin/activate 2>/dev/null
    PYTHON_BIN=$PANEL_PATH/pyenv/bin/python
else
    PYTHON_BIN=/usr/bin/python3
fi

# -----------------------------------------------------------------------------
# start_service
#
# Launches nfs_free_service in the background via nohup. The service script
# calls nfs_free_main.auto_mount(), which iterates over mount.json and mounts
# every entry with auto_mount=1 that is not already mounted.
# Output is discarded because this runs unattended at boot time.
#
# @return  void
# -----------------------------------------------------------------------------
start_service() {
    echo -n "Starting NFS Free auto-mount service... "
    nohup $PYTHON_BIN $SERVICE_SCRIPT > /dev/null 2>&1 &
    echo "done"
}

# -----------------------------------------------------------------------------
# Entry point — SysV action dispatch
# -----------------------------------------------------------------------------
case "$1" in
    start)
        start_service
        ;;
    stop)
        # Auto-mount is a one-shot operation at boot; nothing to stop
        echo "NFS Free service stop requested (auto-mount only runs at boot)"
        ;;
    restart|reload)
        start_service
        ;;
    status)
        echo "NFS Free auto-mount service"
        ;;
    *)
        echo "Usage: /etc/init.d/nfs_free {start|restart|reload|stop|status}"
        exit 1
        ;;
esac


repair.sh
#!/bin/bash
# =============================================================================
#  Network File System (NFS) — Free Edition
#  Repair Script
# =============================================================================
#  Description : Repairs the NFS plugin configuration by re-pinning the
#                mountd, statd, and lockd daemons to their fixed ports in
#                /etc/nfs.conf, then restarting the init service.
#
#                Run this script when:
#                  - NFS mounts fail after a system reboot.
#                  - showmount is unreachable from a remote host.
#                  - mountd is found listening on a random ephemeral port
#                    instead of the expected port 20048.
#
#  Usage       : bash repair.sh
#
#  Exit codes  : 0  success (always; errors are non-fatal)
#
#  Author      : Jose Alexis Correa Valencia (jalexiscv)
#  Email       : jalexiscv@gmail.com
#  GitHub      : https://github.com/jalexiscv
#  LinkedIn    : https://www.linkedin.com/in/jalexiscv/
#  Version     : 1.0
#  License     : MIT — Free Software
#  Date        : July 19, 2023
#  Copyright   : © 2023 Jose Alexis Correa Valencia
# =============================================================================

PLUGIN_PATH=/www/server/panel/plugin/nfs_free

# Source install.sh to reuse fix_mountd_port without re-running the installer.
# The function writes [mountd]/[statd]/[lockd] sections to /etc/nfs.conf and
# restarts nfs-server and rpcbind.
if [ -f "$PLUGIN_PATH/install.sh" ]; then
    source "$PLUGIN_PATH/install.sh" 2>/dev/null
    if declare -f fix_mountd_port >/dev/null 2>&1; then
        fix_mountd_port
    fi
fi

# Restart the auto-mount init service so any pending mounts are retried
/etc/init.d/nfs_free restart 2>/dev/null

echo "Successfully"


upgrade.sh
#!/bin/bash
# =============================================================================
#  Network File System (NFS) — Free Edition
#  Upgrade Script
# =============================================================================
#  Description : Upgrades the NFS system packages to the latest available
#                version from the OS package manager, re-pins the NFS daemon
#                ports via fix_mountd_port, and restarts the init service.
#
#                Run this script when:
#                  - Applying OS security patches that include NFS packages.
#                  - Resolving compatibility issues after a kernel upgrade.
#                  - The panel triggers a plugin upgrade through its interface.
#
#  Usage       : bash upgrade.sh
#
#  Packages upgraded:
#    Debian / Ubuntu  : nfs-kernel-server
#    RHEL / CentOS    : nfs-utils
#
#  Exit codes  : 0  success (always; package manager errors are non-fatal)
#
#  Author      : Jose Alexis Correa Valencia (jalexiscv)
#  Email       : jalexiscv@gmail.com
#  GitHub      : https://github.com/jalexiscv
#  LinkedIn    : https://www.linkedin.com/in/jalexiscv/
#  Version     : 1.0
#  License     : MIT — Free Software
#  Date        : July 23, 2023
#  Copyright   : © 2023 Jose Alexis Correa Valencia
# =============================================================================

PLUGIN_PATH=/www/server/panel/plugin/nfs_free

# Upgrade the NFS package for the detected distribution
if [ -f /usr/bin/apt ]; then
    apt install -y nfs-kernel-server 2>/dev/null
else
    yum install -y nfs-utils 2>/dev/null
fi

# Re-apply fixed port configuration after the package upgrade, since the
# package manager may overwrite /etc/nfs.conf with default values.
if [ -f "$PLUGIN_PATH/install.sh" ]; then
    source "$PLUGIN_PATH/install.sh" 2>/dev/null
    if declare -f fix_mountd_port >/dev/null 2>&1; then
        fix_mountd_port
    fi
fi

# Restart the auto-mount service to pick up any changes
/etc/init.d/nfs_free restart 2>/dev/null

echo "Successfully"


install.sh
#!/bin/bash
# =============================================================================
#  Network File System (NFS) — Free Edition
#  Install / Uninstall Script
# =============================================================================
#  Description : Installs or removes the nfs_free aaPanel plugin.
#                Handles OS package installation, fixed-port configuration
#                for mountd/statd/lockd, firewall rules, SysV init service
#                registration, and systemd service activation.
#
#  Usage       : bash install.sh {install|uninstall}
#
#  Exit codes  : 0  success
#                1  invalid argument
#
#  Author      : Jose Alexis Correa Valencia (jalexiscv)
#  Email       : jalexiscv@gmail.com
#  GitHub      : https://github.com/jalexiscv
#  LinkedIn    : https://www.linkedin.com/in/jalexiscv/
#  Version     : 1.0
#  License     : MIT — Free Software
#  Date        : July 11, 2023
#  Copyright   : © 2023 Jose Alexis Correa Valencia
# =============================================================================

PATH=/www/server/panel/pyenv/bin:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:~/bin
export PATH

PLUGIN_PATH=/www/server/panel/plugin/nfs_free
INIT_SCRIPT=/etc/init.d/nfs_free
SERVICE_NAME=nfs_free

# -----------------------------------------------------------------------------
# install_nfs_packages
#
# Installs the NFS server and client packages required by the plugin.
# Detects the package manager (apt / yum / dnf) and installs the appropriate
# package set. Prints a warning if nfsstat is not found after installation,
# indicating that the packages may not have been installed correctly.
#
# Supported distributions:
#   - Debian / Ubuntu  : nfs-kernel-server, nfs-common
#   - RHEL / CentOS    : nfs-utils  (via yum)
#   - Fedora / Rocky   : nfs-utils  (via dnf)
#
# @return  void
# -----------------------------------------------------------------------------
install_nfs_packages() {
    echo "Installing NFS server packages..."
    if [ -f /usr/bin/apt ]; then
        apt update -qq && apt install -y nfs-kernel-server nfs-common 2>/dev/null
    elif [ -f /usr/bin/yum ]; then
        yum install -y nfs-utils 2>/dev/null
    elif [ -f /usr/bin/dnf ]; then
        dnf install -y nfs-utils 2>/dev/null
    fi

    if [ ! -f /usr/sbin/nfsstat ] && [ ! -f /usr/bin/nfsstat ]; then
        echo "Warning: NFS server packages may not have installed correctly."
        echo "Please install nfs-kernel-server (Debian/Ubuntu) or nfs-utils (RHEL/CentOS) manually."
    fi
}

# -----------------------------------------------------------------------------
# fix_mountd_port
#
# Pins the NFS auxiliary daemons to fixed, well-known ports in /etc/nfs.conf
# so that firewall rules between plugin instances remain stable across reboots.
#
# Ports configured:
#   mountd  → 20048 (TCP/UDP)   — handles mount requests
#   statd   → 32876 (TCP/UDP)   — NSM state recovery
#   lockd   → 32874 (TCP/UDP)   — NLM file locking
#
# Without this configuration mountd uses a random ephemeral port assigned by
# rpcbind at startup, making it impossible to write static firewall rules.
#
# Also opens the required firewall ports in ufw if the service is active.
#
# @return  void
# -----------------------------------------------------------------------------
fix_mountd_port() {
    echo "Configuring NFS mountd on fixed port 20048..."
    NFS_CONF=/etc/nfs.conf
    if [ ! -f "$NFS_CONF" ]; then
        touch "$NFS_CONF"
    fi

    # Add [mountd] section if absent, then set or update port
    if ! grep -q '^\[mountd\]' "$NFS_CONF"; then
        echo "[mountd]" >> "$NFS_CONF"
    fi
    if grep -q '^port=' "$NFS_CONF"; then
        sed -i 's/^port=.*/port=20048/' "$NFS_CONF"
    else
        sed -i '/^\[mountd\]/a port=20048' "$NFS_CONF"
    fi

    # Pin statd port
    if ! grep -q '^\[statd\]' "$NFS_CONF"; then
        echo "[statd]" >> "$NFS_CONF"
        echo "port=32876" >> "$NFS_CONF"
    fi

    # Pin lockd port (TCP and UDP)
    if ! grep -q '^\[lockd\]' "$NFS_CONF"; then
        echo "[lockd]" >> "$NFS_CONF"
        echo "port=32874" >> "$NFS_CONF"
        echo "udp-port=32874" >> "$NFS_CONF"
    fi

    # Open firewall ports when ufw is active
    if command -v ufw >/dev/null 2>&1 && ufw status | grep -q 'active'; then
        for port in 111 2049 20048 32874 32876; do
            ufw allow $port/tcp 2>/dev/null
        done
        echo "Firewall ports opened: 111, 2049, 20048, 32874, 32876"
    fi

    systemctl restart nfs-server 2>/dev/null
    systemctl restart rpcbind 2>/dev/null
    echo "mountd fixed to port 20048"
}

# -----------------------------------------------------------------------------
# install_service
#
# Registers and starts the nfs_free SysV init service that performs
# auto-mounting of configured NFS shares at system boot.
#
# Steps performed:
#   1. Copies init.sh to /etc/init.d/nfs_free and makes it executable.
#   2. Registers the service with update-rc.d (Debian/Ubuntu) or
#      chkconfig (RHEL/CentOS).
#   3. Enables and starts nfs-server and rpcbind via systemd.
#   4. Calls fix_mountd_port to ensure stable port configuration.
#   5. Starts the auto-mount service immediately.
#
# @return  void
# -----------------------------------------------------------------------------
install_service() {
    echo "Setting up auto-mount service..."

    cp -f $PLUGIN_PATH/init.sh $INIT_SCRIPT
    chmod +x $INIT_SCRIPT

    if [ -f /usr/bin/apt-get ]; then
        update-rc.d $SERVICE_NAME defaults 2>/dev/null
    else
        chkconfig --add $SERVICE_NAME 2>/dev/null
        chkconfig --level 2345 $SERVICE_NAME on 2>/dev/null
    fi

    systemctl enable nfs-server 2>/dev/null
    systemctl start nfs-server 2>/dev/null
    systemctl enable rpcbind 2>/dev/null
    systemctl start rpcbind 2>/dev/null

    # Fixed port required for cross-firewall showmount between plugin instances
    fix_mountd_port

    $INIT_SCRIPT start 2>/dev/null
}

# -----------------------------------------------------------------------------
# install_plugin
#
# Full installation entry point. Orchestrates package installation, service
# setup, directory permissions, and panel reload signal.
#
# @return  void
# -----------------------------------------------------------------------------
install_plugin() {
    echo "Installing Network File System (NFS) Free Edition v1.0..."
    echo "Author: Jose Alexis Correa Valencia (jalexiscv)"
    echo ""

    mkdir -p $PLUGIN_PATH/config
    install_nfs_packages
    install_service

    chmod -R 755 $PLUGIN_PATH
    chmod 600 $PLUGIN_PATH/config/*.json 2>/dev/null

    # Signal aaPanel to reload the plugin list
    if [ -f /www/server/panel/data/reload.pl ]; then
        echo > /www/server/panel/data/reload.pl
    fi

    echo ""
    echo "============================================"
    echo " Network File System (NFS) v1.0"
    echo " Installed successfully!"
    echo "============================================"
    echo " Author : Jose Alexis Correa Valencia"
    echo " GitHub : https://github.com/jalexiscv"
    echo " License: MIT — Free Software"
    echo " Date   : July 11, 2023"
    echo "============================================"
    echo ""
}

# -----------------------------------------------------------------------------
# uninstall_plugin
#
# Removes the plugin from the system. Stops and deregisters the init service,
# deletes the plugin directory, and signals the panel to reload.
#
# Note: NFS system packages (nfs-kernel-server / nfs-utils) are intentionally
# preserved so that any manually configured exports continue to work.
# /etc/exports is also left untouched.
#
# @return  void
# -----------------------------------------------------------------------------
uninstall_plugin() {
    echo "Uninstalling Network File System (NFS) Free Edition..."

    if [ -f $INIT_SCRIPT ]; then
        $INIT_SCRIPT stop 2>/dev/null
    fi

    if [ -f /usr/bin/apt-get ]; then
        update-rc.d $SERVICE_NAME remove 2>/dev/null
    else
        chkconfig --del $SERVICE_NAME 2>/dev/null
    fi

    rm -f $INIT_SCRIPT
    rm -rf $PLUGIN_PATH

    if [ -f /www/server/panel/data/reload.pl ]; then
        echo > /www/server/panel/data/reload.pl
    fi

    echo "Network File System (NFS) uninstalled."
    echo "Note: NFS packages (nfs-kernel-server/nfs-utils) were NOT removed."
    echo "Note: /etc/exports was NOT modified."
}

# -----------------------------------------------------------------------------
# Entry point
# -----------------------------------------------------------------------------
case "$1" in
    install)
        install_plugin
        ;;
    uninstall)
        uninstall_plugin
        ;;
    *)
        echo "Usage: $0 {install|uninstall}"
        exit 1
        ;;
esac
